internal/validation/gateway.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/projectcontour/contour/internal/gateway"
+
+	apivalidation "k8s.io/apimachinery/pkg/api/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi_v1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+)
+
+// ValidateName validates that the given name conforms to Kubernetes naming standards.
+var ValidateName = apivalidation.NameIsDNSSubdomain
+
+// ValidateGateway validates gw according to the Gateway API specification.
+// For additional details of the Gateway spec, refer to:
+//   https://gateway-api.sigs.k8s.io/spec/#networking.x-k8s.io/v1alpha1.Gateway
+func ValidateGateway(ctx context.Context, cli client.Client, gw *gatewayapi_v1alpha1.Gateway) field.ErrorList {
+	errs := apivalidation.ValidateObjectMeta(&gw.ObjectMeta, true, ValidateName, field.NewPath("metadata"))
+	errs = append(errs, validateGatewayNamespace(ctx, cli, gw, field.NewPath("metadata"))...)
+	errs = append(errs, validateGatewaySpec(ctx, cli, gw, field.NewPath("spec"))...)
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+// validateGatewayNamespace validates whether other gateways exist in the namespace of gw.
+func validateGatewayNamespace(ctx context.Context, cli client.Client, gw *gatewayapi_v1alpha1.Gateway, path *field.Path) field.ErrorList {
```

Comment:
Right now all validation is done in the DAG builder processor. How do we manage validation here on top of validation there? There are a bunch of bits that could be pulled out and pushed upstream, however, other things, like specific configuration items can't have generic validation and need to set status.

Reply:
IMO validation is one of the first steps that a controller should perform on the object it reconciles. If utilizing the controller-runtime caches is the long-term direction, I think it's best to not add any more functionality to the DAG. As-is, the DAG will continue to perform validation, status mgt, etc. for xRoutes and Gateway listeners. The controller-runtime controllers will do the same for the objects they reconcile. When the xRoute controllers are implemented, we migrate the validation, status mgt, etc. from the Gateway API DAG processor.

Reply:
I wouldn't go down that road right now, I'd let the DAG builder validate its stuff and keep everything together.

To me, this new controller should just manage Envoy instances, everything else should stay consistent as-is.

Reply:
But don't let me distract, if other's don't agree then go with that. =)

Reply:
I think this is a good chance to start pulling functionality out of the DAG, and decomposing it a bit. I think of Contour as having three stages: Ingestion of objects into the DAG, the DAG run, and then export to Envoy. I think that it's okay to have some validation at the ingestion stage, as long as we keep in mind what that means for the DAG. It's just a data model, and if we can split complexity out of it, it will be easier to code and maintain.

Reply:
We'll need to either process a full dag to find some of the validation errors. One simple one is a missing/invalid service reference should set status on the Gateway. Probably for now we'll need to have the Gateway Controller get feedback from the result of a dag run and see if errors exist that it cares about to set proper status.